### ISSUE TITLE
Hessian of custom loss w.r.t parameters

### ISSUE THREAD

**Issue by sanjeevmk** [pytorch/pytorch#52563]
I have an auto-encoder as below (in pseudo-code):

```
code = encoder(x)  # x is input
reconstruction = decoder(code) 
```

I now compute a custom loss on output. This custom loss is a Pytorch extension that I myself wrote. It has both forward and backward defined as illustrated for writing extensions:

```
class CustomLossFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx,reconstruction,x):
         loss = some_operation(reconstruction,x)
         return loss

    @staticmethod
    def backward(ctx,grad_output):
        gradient = some_operation(reconstruction,x,grad_output)
        return gradient

class CustomLoss(torch.nn.Module):
    def forward(self,reconstruction,x):
        return CustomLossFunction.apply(reconstruction,x)
```

Then in my training code:
```
custom_loss = CustomLoss()
code = encoder(x)
reconstruction = decoder(code)
loss = custom_loss(reconstruction,x)
```

Now, I want to compute the Hessian of the ```loss``` wrt the intermediate ```code``` tensor (the bottleneck of the Auto-encoder).

The ```code``` is a tensor of size ```[1,32]``` (a 32-dimensional bottleneck layer).

From other answers written for the same question, I know I can do this:

```
grad = torch.autograd.grad(loss,code,create_graph=True)
# loss is scalar, so grad is same size as code : [1,32]

double_grad = torch.autograd.grad(grad,code,grad_outputs=torch.ones_like(grad))
# expectation is a 32x32 Hessian
```

All of the above runs, but ```double_grad``` is computed as a Zero matrix of 32x32 (all entries zero).

If I replace the ```loss``` with ```nn.MSELoss()``` , then ```double_grad``` gets correct values. But that's not the case for my ```custom_loss```

I believe, that is because, I also need to provide the 2nd derivative of my custom loss w.r.t the ```reconstruction``` (output of decoder). 

But, where do I provide it? I can easily compute the 2nd derivative of my custom operation w.r.t the output of network, and I want to provide something like ```double_backward``` to the ```torch.autograd.Function``` implementation of my operation, so that the second backward operation can use it to compute the Hessian.

Can someone tell me how to solve this? Any method is fine, the above suggestion of ```double_backward``` is simply to explain the problem I'm facing.

Thanks!
